File: src/engine/zobrist.py
# src/engine/zobrist.py

import random
from typing import Optional

import chess


class Zobrist:
    """
    Zobrist hashing implementation with incremental updates.

    Provides efficient position hashing that can be updated incrementally
    as moves are made and unmade, avoiding full board rehashing.
    """

    def __init__(self):
        """Initialize Zobrist hash keys for all board elements."""
        self.keys = {}

        # Piece keys: 12 pieces (6 types x 2 colors) x 64 squares

        for piece in range(1, 7):  # PAWN to KING
            for color in [chess.WHITE, chess.BLACK]:
                for square in range(64):
                    self.keys[(piece, color, square)] = random.randint(0, 2**64 - 1)
        # Castling rights

        self.keys["castling_wk"] = random.randint(0, 2**64 - 1)
        self.keys["castling_wq"] = random.randint(0, 2**64 - 1)
        self.keys["castling_bk"] = random.randint(0, 2**64 - 1)
        self.keys["castling_bq"] = random.randint(0, 2**64 - 1)

        # En passant file

        for file in range(8):
            self.keys[("ep", file)] = random.randint(0, 2**64 - 1)
        # Turn

        self.keys["turn"] = random.randint(0, 2**64 - 1)

        # Current hash value for incremental updates

        self._current_hash: Optional[int] = None

    def hash_board(self, board: chess.Board) -> int:
        """
        Compute the full Zobrist hash for a board position.

        Args:
            board: Chess board to hash

        Returns:
            64-bit Zobrist hash value
        """
        hash_val = 0

        # Hash pieces

        for square in range(64):
            piece = board.piece_at(square)
            if piece:
                hash_val ^= self.keys[(piece.piece_type, piece.color, square)]
        # Hash castling rights

        if board.has_kingside_castling_rights(chess.WHITE):
            hash_val ^= self.keys["castling_wk"]
        if board.has_queenside_castling_rights(chess.WHITE):
            hash_val ^= self.keys["castling_wq"]
        if board.has_kingside_castling_rights(chess.BLACK):
            hash_val ^= self.keys["castling_bk"]
        if board.has_queenside_castling_rights(chess.BLACK):
            hash_val ^= self.keys["castling_bq"]
        # Hash en passant

        if board.ep_square is not None:
            file = chess.square_file(board.ep_square)
            hash_val ^= self.keys[("ep", file)]
        # Hash turn

        if board.turn == chess.BLACK:
            hash_val ^= self.keys["turn"]
        self._current_hash = hash_val
        return hash_val

    def update_hash_for_move(
        self,
        board: chess.Board,
        move: chess.Move,
        old_castling_rights: int,
        old_ep_square: Optional[int],
    ) -> int:
        """
        Incrementally update the hash after a move is made.

        Args:
            board: Board after the move has been made
            move: The move that was made
            old_castling_rights: Castling rights before the move
            old_ep_square: En passant square before the move

        Returns:
            Updated hash value
        """
        if self._current_hash is None:
            return self.hash_board(board)
        hash_val = self._current_hash

        # Toggle turn

        hash_val ^= self.keys["turn"]

        # Handle piece movement

        from_square = move.from_square
        to_square = move.to_square

        # Get the piece that moved (it's now on the to_square)

        moved_piece = board.piece_at(to_square)
        if moved_piece:
            # Remove piece from old square

            hash_val ^= self.keys[
                (moved_piece.piece_type, moved_piece.color, from_square)
            ]
            # Add piece to new square

            hash_val ^= self.keys[
                (moved_piece.piece_type, moved_piece.color, to_square)
            ]
        # Handle captures (remove captured piece)

        if board.is_capture(move):
            if board.is_en_passant(move):
                # En passant capture - remove the captured pawn

                if (
                    board.turn == chess.WHITE
                ):  # White just moved, so black pawn was captured
                    captured_square = to_square - 8
                    hash_val ^= self.keys[(chess.PAWN, chess.BLACK, captured_square)]
                else:  # Black just moved, so white pawn was captured
                    captured_square = to_square + 8
                    hash_val ^= self.keys[(chess.PAWN, chess.WHITE, captured_square)]
            else:
                # Regular capture - we need to determine what was captured
                # Since the move is already made, we need to infer the captured piece
                # This is a limitation - ideally we'd pass the captured piece info

                pass  # For now, fall back to full rehash for captures
        # Handle promotion

        if move.promotion:
            # Remove the pawn that was promoted

            hash_val ^= self.keys[(chess.PAWN, moved_piece.color, to_square)]
            # The promoted piece is already added above, so we're good
        # Handle castling

        if board.is_castling(move):
            # Handle rook movement in castling

            if move.to_square == chess.G1:  # White kingside
                hash_val ^= self.keys[(chess.ROOK, chess.WHITE, chess.H1)]
                hash_val ^= self.keys[(chess.ROOK, chess.WHITE, chess.F1)]
            elif move.to_square == chess.C1:  # White queenside
                hash_val ^= self.keys[(chess.ROOK, chess.WHITE, chess.A1)]
                hash_val ^= self.keys[(chess.ROOK, chess.WHITE, chess.D1)]
            elif move.to_square == chess.G8:  # Black kingside
                hash_val ^= self.keys[(chess.ROOK, chess.BLACK, chess.H8)]
                hash_val ^= self.keys[(chess.ROOK, chess.BLACK, chess.F8)]
            elif move.to_square == chess.C8:  # Black queenside
                hash_val ^= self.keys[(chess.ROOK, chess.BLACK, chess.A8)]
                hash_val ^= self.keys[(chess.ROOK, chess.BLACK, chess.D8)]
        # Update castling rights

        self._update_castling_hash(hash_val, old_castling_rights, board.castling_rights)

        # Update en passant

        if old_ep_square is not None:
            old_file = chess.square_file(old_ep_square)
            hash_val ^= self.keys[("ep", old_file)]
        if board.ep_square is not None:
            new_file = chess.square_file(board.ep_square)
            hash_val ^= self.keys[("ep", new_file)]
        self._current_hash = hash_val
        return hash_val

    def _update_castling_hash(
        self, hash_val: int, old_rights: int, new_rights: int
    ) -> None:
        """Update hash for castling rights changes."""
        # XOR out old rights and XOR in new rights

        if (old_rights & chess.BB_H1) != (new_rights & chess.BB_H1):
            hash_val ^= self.keys["castling_wk"]
        if (old_rights & chess.BB_A1) != (new_rights & chess.BB_A1):
            hash_val ^= self.keys["castling_wq"]
        if (old_rights & chess.BB_H8) != (new_rights & chess.BB_H8):
            hash_val ^= self.keys["castling_bk"]
        if (old_rights & chess.BB_A8) != (new_rights & chess.BB_A8):
            hash_val ^= self.keys["castling_bq"]

    def get_current_hash(self) -> Optional[int]:
        """Get the current hash value without recalculating."""
        return self._current_hash

    def invalidate_hash(self) -> None:
        """Invalidate the current hash, forcing recalculation on next access."""
        self._current_hash = None

File: src/engine/_cpp/search/zobrist.cpp
// ---------------------------------------------------------------------------
// zobrist.cpp — High-performance Zobrist hashing (non-PolyGlot)
//
// Key design choices vs. the old PolyGlot implementation:
//   1. Compile-time SplitMix64 keys in zobrist_keys.hpp — zero init cost.
//   2. piece[color][pt][sq] direct 3-D indexing — no offset arithmetic.
//   3. castling[rights_mask] single-lookup 16-entry table — replaces 4
//      conditional XORs.
//   4. EP hashed unconditionally by file — no pawn-can-capture filter,
//      which was a PolyGlot spec detail irrelevant for engine use.
//   5. make_null_move_hash() — O(1) null-move hash (toggle side + remove
//      old EP).
//   6. [[unlikely]] on cold fallback paths for branch-prediction hints.
// ---------------------------------------------------------------------------

#include "zobrist.hpp"

#include "zobrist_keys.hpp"

namespace search {

Zobrist::Zobrist(std::optional<uint64_t> /*seed*/) noexcept
    : current_hash_(std::nullopt) {}

// ---------------------------------------------------------------------------
// hash_board — Full O(popcount) hash from scratch
// ---------------------------------------------------------------------------
uint64_t Zobrist::hash_board(const Board &board) noexcept {
  uint64_t h = 0;

  // 1. Pieces — iterate bitboards with hardware CTZ/BLSR
  for (uint8_t pt = 0; pt < NUM_PIECE_TYPES; ++pt) {
    for (uint8_t c = 0; c < NUM_COLORS; ++c) {
      Bitboard bb =
          board.get_piece_bb(static_cast<PieceType>(pt), static_cast<Color>(c));
      while (bb) {
        h ^= ZKEYS.piece[c][pt][pop_lsb(bb)];
      }
    }
  }

  // 2. Castling — single indexed load (no per-bit branching)
  h ^= ZKEYS.castling[board.get_castling_rights()];

  // 3. En passant — hash file unconditionally when EP is set
  const int8_t ep = board.get_en_passant_square();
  if (ep != -1) {
    h ^= ZKEYS.ep_file[square_file(static_cast<uint8_t>(ep))];
  }

  // 4. Side to move — XOR when white to move
  if (board.get_side_to_move()) {
    h ^= ZKEYS.side;
  }

  current_hash_ = h;
  return h;
}

// ---------------------------------------------------------------------------
// make_move_hash — O(1) incremental update
//
// Given the current hash and a pre-move Board, computes the hash of the
// position that results from applying 'move'.  Uses the Board's mailbox
// arrays for O(1) piece identification.
// ---------------------------------------------------------------------------
uint64_t Zobrist::make_move_hash(Board &board, const Move &move) noexcept {
  // Cold fallback: no hash cached yet.
  if (!current_hash_.has_value()) [[unlikely]] {
    board.push(move);
    const uint64_t h = hash_board(board);
    (void)board.pop();
    return h;
  }

  uint64_t h = current_hash_.value();

  const uint8_t from = move.from;
  const uint8_t to = move.to;
  const uint8_t promo = move.promotion;

  // Identify moving piece via mailbox — O(1)
  const auto piece_opt = board.piece_at(from);
  if (!piece_opt.has_value()) [[unlikely]] {
    board.push(move);
    const uint64_t fh = hash_board(board);
    (void)board.pop();
    return fh;
  }

  const uint8_t moving_pt = static_cast<uint8_t>(piece_opt->type);
  const uint8_t us = static_cast<uint8_t>(piece_opt->color);
  const uint8_t them = us ^ 1u;

  const int8_t old_ep = board.get_en_passant_square();
  const uint8_t old_cr = board.get_castling_rights();

  // --- 1. Toggle side to move (always flips) ---
  h ^= ZKEYS.side;

  // --- 2. Remove old en-passant contribution ---
  if (old_ep != -1) {
    h ^= ZKEYS.ep_file[square_file(static_cast<uint8_t>(old_ep))];
  }

  // --- 3. Remove old castling contribution (single lookup) ---
  h ^= ZKEYS.castling[old_cr];

  // --- 4. XOR out moving piece from source ---
  h ^= ZKEYS.piece[us][moving_pt][from];

  // --- 5. Handle captures ---
  bool is_ep = false;
  bool is_capt = false;
  uint8_t capt_sq = to;

  // Check en-passant capture
  if (moving_pt == static_cast<uint8_t>(PieceType::PAWN) && old_ep != -1 &&
      to == static_cast<uint8_t>(old_ep)) {
    const auto target_piece = board.piece_at(to);
    if (!target_piece.has_value() || !target_piece->valid) {
      is_ep = true;
      is_capt = true;
      capt_sq = static_cast<uint8_t>(old_ep + (us == 0u ? -8 : 8));
    }
  }

  if (is_ep) {
    // XOR out the captured pawn
    h ^= ZKEYS.piece[them][static_cast<uint8_t>(PieceType::PAWN)][capt_sq];
  } else {
    // Normal capture — check mailbox at destination
    const auto captured = board.piece_at(to);
    if (captured.has_value() && captured->valid) {
      is_capt = true;
      h ^= ZKEYS.piece[static_cast<uint8_t>(captured->color)]
                      [static_cast<uint8_t>(captured->type)][to];
    }
  }

  // --- 6. XOR in the piece at destination ---
  if (promo != 0) {
    // Promotion: arriving piece is the promoted type
    h ^= ZKEYS.piece[us][promo][to];
  } else {
    h ^= ZKEYS.piece[us][moving_pt][to];
  }

  // --- 7. Handle castling rook movement ---
  const int delta = static_cast<int>(to) - static_cast<int>(from);
  const bool is_castling =
      (moving_pt == static_cast<uint8_t>(PieceType::KING)) &&
      (delta == 2 || delta == -2);

  if (is_castling) {
    const bool kingside = (to > from);
    const uint8_t rook_from =
        kingside ? (us == 0u ? 7u : 63u) : (us == 0u ? 0u : 56u);
    const uint8_t rook_to = kingside ? static_cast<uint8_t>(from + 1u)
                                     : static_cast<uint8_t>(from - 1u);
    constexpr uint8_t ROOK_PT = static_cast<uint8_t>(PieceType::ROOK);
    h ^= ZKEYS.piece[us][ROOK_PT][rook_from];
    h ^= ZKEYS.piece[us][ROOK_PT][rook_to];
  }

  // --- 8. Compute new castling rights ---
  uint8_t new_cr = old_cr;
  if (new_cr) {
    if (moving_pt == static_cast<uint8_t>(PieceType::KING)) {
      new_cr &= static_cast<uint8_t>(us == 0u ? ~0x03u : ~0x0Cu);
    }
    // Rook leaving its home corner or captured on its home corner
    auto revoke = [&](uint8_t sq, uint8_t mask) noexcept {
      if (from == sq || (is_capt && capt_sq == sq))
        new_cr &= static_cast<uint8_t>(~mask);
    };
    revoke(0u, 0x02u);  // a1 — white queenside
    revoke(7u, 0x01u);  // h1 — white kingside
    revoke(56u, 0x08u); // a8 — black queenside
    revoke(63u, 0x04u); // h8 — black kingside
  }
  h ^= ZKEYS.castling[new_cr];

  // --- 9. Compute new en-passant square ---
  // Double pawn push → EP square; otherwise cleared.
  if (moving_pt == static_cast<uint8_t>(PieceType::PAWN) &&
      (delta == 16 || delta == -16)) {
    const auto new_ep = static_cast<uint8_t>(from + (us == 0u ? 8 : -8));
    h ^= ZKEYS.ep_file[square_file(new_ep)];
  }

  return h;
}

// ---------------------------------------------------------------------------
// make_null_move_hash — O(1) incremental null-move update
//
// A null move only flips side-to-move and clears the EP square.  No piece,
// castling, or other state changes.  This avoids the full O(popcount)
// hash_board() call that the old code used after push_null().
// ---------------------------------------------------------------------------
uint64_t Zobrist::make_null_move_hash(const Board &board) noexcept {
  if (!current_hash_.has_value()) [[unlikely]] {
    // No hash available — compute from scratch for current position,
    // then apply the null-move delta.
    // (This cast is safe: hash_board only reads the board.)
    uint64_t h = hash_board(const_cast<Board &>(
        board)); // NOLINT(cppcoreguidelines-pro-type-const-cast)
    // Now apply null-move: toggle side, remove EP.
    h ^= ZKEYS.side;
    const int8_t ep = board.get_en_passant_square();
    if (ep != -1) {
      h ^= ZKEYS.ep_file[square_file(static_cast<uint8_t>(ep))];
    }
    return h;
  }

  uint64_t h = current_hash_.value();

  // Toggle side to move
  h ^= ZKEYS.side;

  // Remove old EP contribution (null move clears EP — no new EP to add)
  const int8_t ep = board.get_en_passant_square();
  if (ep != -1) {
    h ^= ZKEYS.ep_file[square_file(static_cast<uint8_t>(ep))];
  }

  return h;
}

// ---------------------------------------------------------------------------
// Accessors / mutators
// ---------------------------------------------------------------------------

std::optional<uint64_t> Zobrist::get_current_hash() const noexcept {
  return current_hash_;
}

void Zobrist::set_current_hash(std::optional<uint64_t> hash_val) noexcept {
  current_hash_ = hash_val;
}

void Zobrist::invalidate_hash() noexcept { current_hash_ = std::nullopt; }

} // namespace search
